### ISSUE TITLE
invalid memory address or nil pointer when running with --tags

### ISSUE THREAD

**Issue by AdamJorgenson** [getgauge/gauge#1364]
### Expected behavior
Should be able to run specs with tags

### Actual behavior
Cannot run specs with --tags argument

### Steps to reproduce
1.On an OSX 10.13 machine try gauge run --tags TT-Group1 specs/ we also tried gauge run --tags=TT-Group1 specs/ and "gauge run --tags="TT-Group1" specs/
2. If the command  is "gauge run specs/somespec.spec" it works fine

Error ----------------------------------
Panicing : runtime error: invalid memory address or nil pointer dereference
goroutine 1 [running]:
runtime/debug.Stack(0xc0005ab6c0, 0x166e760, 0x1cabc40)
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0xa7
main.recoverPanic()
	src/github.com/getgauge/gauge/gauge.go:36 +0x57
panic(0x166e760, 0x1cabc40)
	/usr/local/opt/go/libexec/src/runtime/panic.go:513 +0x1b9
github.com/getgauge/gauge/filter.filterSpecsByTags(0xc000486800, 0x8b7, 0x8b7, 0x7ffeefbffab9, 0x9, 0xc0004a8000, 0x1, 0x1e89000)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/filter/specItemFilter.go:214 +0xcd
github.com/getgauge/gauge/filter.(*tagsFilter).filter(0xc00fb2ea10, 0xc000486800, 0x8b7, 0x8b7, 0x100dc78, 0x20, 0x1687fe0)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/filter/specsFilter.go:45 +0xa3
github.com/getgauge/gauge/filter.applyFilters(0xc000486800, 0x8b7, 0x8b7, 0xc0005ab9c0, 0x3, 0x3, 0x900, 0x10000c00019da00, 0x314)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/filter/filter.go:31 +0x89
github.com/getgauge/gauge/filter.FilterSpecs(0xc000486800, 0x8b7, 0x8b7, 0x3, 0xc0005730a0, 0xc000486800)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/filter/filter.go:15 +0x19c
github.com/getgauge/gauge/parser.ParseSpecs(0xc000187110, 0x1, 0x3, 0xc0003e0060, 0xc0005730a0, 0xc0000effa0, 0x1c, 0x1740f9a, 0xb)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/parser/parse.go:87 +0x85
github.com/getgauge/gauge/validation.ValidateSpecs(0xc000187110, 0x1, 0x3, 0x0, 0x0)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/validation/validate.go:175 +0x120
github.com/getgauge/gauge/execution.glob..func1(0xc000187110, 0x1, 0x3, 0x0)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/execution/execute.go:131 +0xb2
github.com/getgauge/gauge/cmd.execute(0x1cb5ca0, 0xc000187110, 0x1, 0x3)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/cmd/run.go:205 +0x102
github.com/getgauge/gauge/cmd.glob..func14(0x1cb5ca0, 0xc000187110, 0x1, 0x3)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/cmd/run.go:93 +0xfb
github.com/getgauge/gauge/vendor/github.com/spf13/cobra.(*Command).execute(0x1cb5ca0, 0xc000187080, 0x3, 0x3, 0x1cb5ca0, 0xc000187080)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/vendor/github.com/spf13/cobra/command.go:651 +0x24a
github.com/getgauge/gauge/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1cb6fc0, 0xf, 0xc000192180, 0xc0001424b0)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/vendor/github.com/spf13/cobra/command.go:726 +0x2cc
github.com/getgauge/gauge/vendor/github.com/spf13/cobra.(*Command).Execute(0x1cb6fc0, 0x0, 0xc00019df78)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/vendor/github.com/spf13/cobra/command.go:685 +0x2b
github.com/getgauge/gauge/cmd.Parse(0x0, 0x176c960)
	/private/tmp/gauge-20190125-2013-18yzinh/gauge-1.0.4/src/github.com/getgauge/gauge/cmd/cmd.go:154 +0x3d
main.main()
	src/github.com/getgauge/gauge/gauge.go:29 +0x3e

	

### Gauge version
```
darwin, 1.0.4
	html-report (4.0.6), java (0.7.1), screenshot (0.0.1), xml-report (0.2.1)
```

**Comment by nehashri** [comment]
@AdamJorgenson This seems to be working fine for us. Can you share a sample project to replicate this issue.

**Comment by ocary** [comment]
@nehashri - It looks like it is something in some our projects.   I did a _gauge java init_ then made a couple more really simple specs with tags in that project.  Running with --tags from the terminal is executing as expected for the sample init project.

When running our project, the error message is as @AdamJorgenson said above, but also there are many ParseErrors right away.
`[ParseError] /...directorystuff/specname.spec:0 Failed to read the file /...directorystuff/specname.spec. => ''`

There are many ParseErrors like this, and the specname.spec file does exist, and the whole project folder has full permissions to rule that out (chmod -R 777 on the base gauge project directory).

I found one of our Gauge projects that this issue does not happen upon (tags from terminal appears to be working as expected), so it is almost certainly something specific to the project.  I'll research more on Monday.

**Comment by ocary** [comment]
@nehashri - It is a scalability issue.   

# ReproSteps
Using macOS Mojave
1. Create a new directory
1. Run _gauge init java_ in that directory
1. In the specs directory, run _for i in {1..1500}; do cp example.spec "example$i.spec"; done_
1. In the main project directory created in the first step, run _gauge run --tags MyTag specs/_
1. Observe errors:   _[ParseError] /..path../example[number].spec: 0 Failed to read the file /..path..example[number].spec. => ''_    and _Panicing : runtime error: invalid memory address or nil pointer deference_

^ If the above does not produce the error, try again with a value larger than 1500.

On one of our projects I've seen the above behavior intermittently when the number of specs is just right.  Always orks as expected at 300 specs.  Intermittent errors at 650 specs.  Always errors at 1000 specs.   <- these numbers seem very project dependent.

**Comment by NivedhaSenthil** [comment]
Tried replicating the above steps with sample java project but no luck reproducing the error. 
I could run even 8000 specs with tags without any issue. Seems to be something else that is causing the issue.

Environment used:
mocOS Mojave
Gauge version: 1.0.4
Commit Hash: 3a9a647

Plugins
-------
html-report (4.0.8.nightly-2019-03-08)
java (0.7.1)
screenshot (0.0.1)

**Comment by ocary** [comment]
Following the repro steps, I've seen this issue on both mac machines I've tried this upon.  One was a macbook and the other an iMac.    On the iMac, I used 4500 specs..

Next week I'll have access to a different mac and will try on there as well, unless someone else is able to reproduce this issue before then.

**Comment by Apoorva-GA** [comment]
@ocary Any luck in reproducing the issue?

**Comment by ocary** [comment]
@Apoorva-GA / @nehashri  - Sorry for the delay.

Yes, I was just able to reproduce this on a basically clean iMac.  A personal machine; no company software that could be causing problems.

Steps done, following the GetGauge.io -> Getting Started page.

1. Installed Node.JS 10.15.3
1. Did a _sudo chown -R......_ to get so I could install gauge; permissions denied on install without doing this.   https://stackoverflow.com/questions/47252451/permission-denied-when-installing-npm-modules-in-osx
1. Installed Gauge 1.0.4 via _npm install -g @getgauge/cli_
1. Made a new directory.
1. Ran _gauge init java_ in that directory
1. Installed Java JDK 11.0.2, after trying _gauge run specs_ and realizing no JDK installed
1. Ran _gauge run specs_ and saw that _example.spec_ ran as expected.
1. in the specs directory, ran _for i in {1..5000}; do cp example.spec "example$i.spec"; done_
1. In the specs parent directory, ran _gauge run specs_ and got the _Failed to read the file // invalid memory address_ errors.

After doing the above, I also tried using 1500 specs instead of 5000, and get the same errors.

This mac is late 2013 iMac.   3.2GHz Intel Core i5.  8GB RAM.   OS = macOS High Sierra.

JDK 11.0.2

Gauge 1.0.4
Java (0.7.1)
html-report (4.0.8)
screenshot (0.0.1)

**Comment by shubhamsc** [comment]
@ocary @AdamJorgenson Tried to replicate this issue.
I get the error only once which @AdamJorgenson mention above https://github.com/getgauge/gauge/issues/1364#issue-418488458. After that, i tried many times but didn't get any error. I followed these steps
>Made a new directory.
Ran gauge init java in that directory
Ran gauge run specs and saw that example.spec ran as expected.
in the specs directory, ran for i in {1..5000}; do cp example.spec "example$i.spec"; done
I also tried using 1500 specs instead of 5000

mac details: -
MacBook Pro 2018,  2.2 GHz Intel Core i7,  16GB macOS Mojave.

All tests are running fine but I'm getting another error with HTML report.
```
Plugin [Html Report] with pid [11001] did not exit after 4.00 seconds. Forcefully killing it.
Specifications:	5001 executed	5001 passed	0 failed	0 skipped
Scenarios:	5001 executed	5001 passed	0 failed	0 skipped
```

**Comment by ocary** [comment]
@shubhamsc / @nehashri - when you are trying to reproduce this, are you running from an actual terminal window, or in a terminal prompt within an IDE?

Oddly, one machine here on a fresh project with 18,000 specs, running from an actual terminal gives the error, but exact same command run in IntelliJ Community 2018.3 runs as expected.   I believe the error was seen once and only once in IntelliJ on this fresh project.

**Comment by BugDiver** [comment]
@ocary I tried the same steps but no luck reproducing this.
> when you are trying to reproduce this, are you running from an actual terminal window, or in a terminal prompt within an IDE?

I used the actual terminal (i-term2 with `zsh` shell).

**Comment by ocary** [comment]
Progress!  All signs point to this being a _open files limit_ problem.

By default _ulimit -Sn_ shows 256   <-- Running specs here gives the error.

Changing that to 4096 via _ulimit -Sn 4096_, then running specs works as expected.


Interesting side note:  The terminal inside IntelliJ 2018.3.3 CE, has on open files value of 12544.   All other values shown in _ulimit -a_ are the same between IntelliJ terminal and a "normal" terminal window.

**Comment by nehashri** [comment]
@ocary Can you give us a little more info, was this reproducible with limit of 256 when running just `gauge run specs` or did you have to run it with specific tags?

**Comment by ocary** [comment]
It turns out this has (likely) nothing to do with _tags_.   The key to getting the error is a _small_ value for number of open files as reported by ```ulimit -n```.    Below are details on versions used, getting Gauge to error and getting specs to run as expected.   Changing the ```ulimit -n``` value has been seen to get Gauge working/breaking on multiple different OS X machines.

# Gauge version details
```
Gauge version: 1.0.4

Plugins
-------
html-report (4.0.8)
java (0.7.2)
screenshot (0.0.1)
xml-report (0.2.1)
```

# Machine used for these reproduction steps
MacBook Pro - Mojave.   16GB memory

# Repro instructions
**all** of this is a standard terminal window

## Initial setup and getting specs to error
* running ```ulimit -n```   gives output ```256``` .  <<---   **This value being small is important**

Create a new java project with many specs.   1500 is used here.
1. Create a new directory
1. Run ```gauge init java``` in the new directory
1. In the specs directory, run ```for i in {1..1500}; do cp example.spec "example$i.spec"; done```
1. In the main project directory created in the first step, run ```gauge run specs/```
1. Observe errors: _[ParseError] /..path../example[number].spec: 0 Failed to read the file /..path..example[number].spec. => '' and Panicing : runtime error: invalid memory address or nil pointer dereference_

## Getting specs to run as expected
This assumes all of the above steps in _Initial setup..._ above have been completed

1. Run ```ulimit -n 4096```    <-- this can be any _large_ number.  IntelliJ terminal uses 10240
1. Run ```ulimit -n``` to very the output is the _large_ number from the above step
1. Run ```gauge run specs/```
1. Observe specs run as expected.



## Reference 

### Terminal -ulimit -a output
```
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) [256 | 4096]       < -----    **This is the value that gets changed**
pipe size            (512 bytes, -p) 1
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 1418
virtual memory          (kbytes, -v) unlimited
```

**Comment by BugDiver** [comment]
@ocary We really appreciate all the good work you did.

With this info, we are able to replicate the issue.

The issue happens when we try to read/parse spec files. The parsing happens concurrently using goroutines. The number of goroutines is equal to the number of specifications. So when the user has a lower rLimit (max no of open file descriptors) some goroutines are not able to read the file.

We are exploring the following solutions.
* Start only rLimit / 2 goroutines,
* Start only 20 goroutines if fail to get the rLimit

We will be using Lazy strategy for parallelization.

Once again Thanks for your time.

**Comment by Debashis9012** [comment]
This has been verified and found fixed with the version ``` Gauge version: 1.0.6.nightly-2019-05-17 ```
